lib/Backend/BackendUtils.cpp

```diff
@@ -123,6 +125,108 @@ bool isInput(const Placeholder *PH) {
+/// Packed into {Placeholder *, isInput, isOutput}.
+std::vector<std::tuple<Placeholder const *, bool, bool>>
+getContiguousPlaceHolder(std::list<Placeholder *> const &holders) {
+  // Pure input placeholders.
+  std::vector<Placeholder const *> intputPlaceholders;
+  // Pure output placeholders.
+  std::vector<Placeholder const *> outputPlaceholders;
+  // Input&output placeholders.
+  std::vector<Placeholder const *> inputOutputPlaceholders;
+  // Neither input nor output placeholders.
+  std::vector<Placeholder const *> emptyPlaceholders;
+  // return value.
+  std::vector<std::tuple<Placeholder const *, bool, bool>> ret;
+
+  for (auto &v : holders) {
+    if (isInput(v)) {
+      if (!isOutput(v)) {
+        intputPlaceholders.push_back(v);
+      } else {
+        inputOutputPlaceholders.push_back(v);
+      }
+    } else {
+      if (isOutput(v)) {
+        outputPlaceholders.push_back(v);
+      } else {
+        emptyPlaceholders.push_back(v);
+      }
+    }
+  }
+
+  for (auto &v : intputPlaceholders) {
+    ret.push_back(std::make_tuple(v, true, false));
+  }
+
+  for (auto &v : inputOutputPlaceholders) {
+    ret.push_back(std::make_tuple(v, true, true));
+  }
+
+  for (auto &v : outputPlaceholders) {
+    ret.push_back(std::make_tuple(v, false, true));
+  }
+  // Currently, we have to allocate this to pass BackendCorrectnessTest.
+  for (auto &v : emptyPlaceholders) {
+    ret.push_back(std::make_tuple(v, false, false));
+  }
+
+  return ret;
+}
+
+/// Convert placeholders to be ordered as input|inputOutput|output|neither.
+/// Packed into {Placeholder *, isInput, isOutput}.
+std::vector<std::tuple<Placeholder const *, bool, bool>>
+getContiguousPlaceHolder(std::vector<Placeholder const *> const &holders) {
+  // Pure input placeholders.
+  std::vector<Placeholder const *> intputPlaceholders;
+  // Pure output placeholders.
+  std::vector<Placeholder const *> outputPlaceholders;
+  // Input&output placeholders.
+  std::vector<Placeholder const *> inputOutputPlaceholders;
+  // Neither input nor output placeholders.
+  std::vector<Placeholder const *> emptyPlaceholders;
+  // return value.
+  std::vector<std::tuple<Placeholder const *, bool, bool>> ret;
+
+  for (auto &v : holders) {
+    if (isInput(v)) {
+      if (!isOutput(v)) {
+        intputPlaceholders.push_back(v);
+      } else {
+        inputOutputPlaceholders.push_back(v);
+      }
+    } else {
+      if (isOutput(v)) {
+        outputPlaceholders.push_back(v);
+      } else {
+        emptyPlaceholders.push_back(v);
+      }
+    }
+  }
+
+  for (auto &v : intputPlaceholders) {
+    ret.push_back(std::make_tuple(v, true, false));
+  }
+
+  for (auto &v : inputOutputPlaceholders) {
+    ret.push_back(std::make_tuple(v, true, true));
+  }
+
+  for (auto &v : outputPlaceholders) {
+    ret.push_back(std::make_tuple(v, false, true));
+  }
+  // Currently, we have to allocate this to pass BackendCorrectnessTest.
```

Comment:
> BackendCorrectnessTest

do you understand why test fails? (is the test correct?)

Reply:
> > BackendCorrectnessTest
> 
> do you understand why test fails? (is the test correct?)

It falls in CPU/CPUOnly.dataParallelStackingTest and report a segmentation fault:11 error. Haven't figured out why currently. 
![Screen Shot 2019-06-07 at 1 50 04 PM](https://user-images.githubusercontent.com/50883732/59133138-ee87ff80-892b-11e9-9184-2d9a2f6ac777.png)